### PR TITLE
[JBTM-3383] Improve ARJUNA016009: Caught::java.lang.NullPointerException log message

### PR DESCRIPTION
**Pull request by ivassile**
Issue: https://issues.redhat.com/browse/JBTM-3383

MAIN AS_TESTS QA_JTA
!QA_JTS_JDKORB !QA_JTS_OPENJDKORB !QA_JTS_JACORB !BLACKTIE !XTS !PERF NO_WIN !RTS !TOMCAT !JACOCO !LRA

### CODE REVIEW

**Comment by ivassile** [comment]
Note: We need to review existing tests. After the change test [2] will fail, because it uses xid with globalTransactionId = null.

[2]
[ERROR] Failures:
[ERROR] XARecoveryModuleUnitTest.testXAResourceRecoveryHelperDeregisterLocking:278 Expecting to get the resource registered by helper because the first pass was run just now expected:<com.hp.mwtests.ts.jta.recovery.TestXAResource@3b2cf7ab> but was:<null>

**Comment by ochaloup** [comment]
TESTIT

File: ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java [hunk 1]
```diff
@@ -736,33 +738,29 @@ private final void xaRecoveryFirstPass(NameScopedXAResource xares)
...
-			}
+
+                if (trans != null) {
+                    List<Xid> transList = new ArrayList<Xid>(Arrays.asList(trans));
+                    Iterator<Xid> it = transList.iterator();
+                    while (it.hasNext()) {
+                        Xid xid = it.next();
+                        if (xid != null) {
+                            byte[] globalTransactionId = xid.getGlobalTransactionId();
+                            if (globalTransactionId != null) {
```
Comment (ochaloup) [inline ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java:749]: A topic for discussion - would make sense to check for `null` the format and branch as well? Is there some experience that e.g. Oracle could not be filling the other parts of Xid and leaving them `null`?

File: ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java [hunk 2]
```diff
@@ -736,33 +738,29 @@ private final void xaRecoveryFirstPass(NameScopedXAResource xares)
...
+                                if (jtaLogger.logger.isDebugEnabled()) {
+                                    jtaLogger.logger.debug("Recovered: "
```
Comment (ochaloup) [inline ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java:751]: A code style thing (not necessary to be addressed) - maybe the `debug` call message could be whole at single line.

File: ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java [hunk 3]
```diff
@@ -1223,6 +1221,40 @@ private void saveContactedJndiName(final String jndiName) {
...
     }
+    
+    private String getXidLogInfo (Xid xid) {
+        StringBuilder stringBuilder = new StringBuilder();
+        
+        if (xid != null) {
+            byte[] globalTransactionId = xid.getGlobalTransactionId();
+            byte[] branchQualifier = xid.getBranchQualifier();
+            stringBuilder.append("< ");
+            stringBuilder.append(xid.getFormatId());
```
Comment (ochaloup) [inline ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java:1232]: If there is the check for `globalTransactionId` and `branchQualifier` in this printing, it could be good to check the format for `null` as well here.

**Review by ochaloup — CHANGES_REQUESTED** [review]
I have few requests for adjustments but nothing crucial. The idea as whole seems good to me.

**Review by ochaloup — CHANGES_REQUESTED** [review]
@ivassile I'm sorry for changing my opinion but I verified the XA spec and JTA spec and I think my consideration was wrong last time.
The spec does not require the global part of `Xid` has not to be null. Such Xid should not be removed from recovery.

Details from XA spec

```
Although "xa.h" constrains the length and byte alignment of the data element within an
XID, it does not specify the data’s contents. The only requirement is that both gtrid and
bqual, taken together, must be globally unique.
```
I would ask you for changing this PR in  this regard - ie. do not remove the null global id from the `trans` array. Please, change only the logging machinery to not throwing the NPE when it's null (either global, or branch, or format).

**Comment by ivassile** [comment]
@ochaloup Thank you for the review. I'll analyze the new requirements and rework the PR.

**Comment by mmusgrov** [comment]
> Note: We need to review existing tests. After the change test [2] will fail, because it uses xid with globalTransactionId = null.
> 
> [2]
> 
> ```
> [ERROR] Failures:
> [ERROR] XARecoveryModuleUnitTest.testXAResourceRecoveryHelperDeregisterLocking:278 Expecting to get the resource registered by helper because the first pass was run just now expected:<com.hp.mwtests.ts.jta.recovery.TestXAResource@3b2cf7ab> but was:<null>
> ```

Is this something you need us to help you with?

File: ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java [hunk 4]
```diff
@@ -736,33 +738,29 @@ private final void xaRecoveryFirstPass(NameScopedXAResource xares)
...
+                        if (xid != null) {
+                            byte[] globalTransactionId = xid.getGlobalTransactionId();
+                            if (globalTransactionId != null) {
+                                if (jtaLogger.logger.isDebugEnabled()) {
+                                    jtaLogger.logger.debug("Recovered: "
+                                        + getXidLogInfo(xid));
+                                }
+                            } else {
+                                it.remove();
+                                jtaLogger.logger.warn("Failed to recover: Returned global transaction identifier is null. "
```
[the remark this answers is not in the document]
Comment (ivassile) [reply ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java]: Resolved

File: ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java [hunk 5]
```diff
@@ -736,33 +738,29 @@ private final void xaRecoveryFirstPass(NameScopedXAResource xares)
...
+                                    + xares.toString() + getXidLogInfo(xid));
+                            }   
+                        }                                  
+                    }
+                    trans = transList.toArray(new Xid[transList.size()]);
```
[the remark this answers is not in the document]
Comment (ivassile) [reply ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java]: Resolved

File: ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java (hunk 2 above)
Comment (ivassile) [reply ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java]: Resolved

File: ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java (hunk 3 above)
Comment (ivassile) [reply ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java:1231]: xid.getFormatId() is int

File: ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java (hunk 1 above)
Comment (ivassile) [reply ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java]: I added check for branch == null.

**Comment by ivassile** [comment]
@ochaloup I corrected most of the requested changes. Based on [1], I think we still need to reject and not process transactions where gtrid and bqual are both null, because the combine value won't be unique.
I also modified XidImple to handle gtrid null value.

@mmusgrov As we discussed, it will be great if you can help with creating a new test which will simulate Oracle(not instanceof XidImple) transaction to replicate the NPE before this change. If we decide to keep my current suggestion and remove transactions where gtrid and bqual are both null, we also need to modify the existing test testXAResourceRecoveryHelperDeregisterLocking.

[1] Although "xa.h" constrains the length and byte alignment of the data element within an
XID, it does not specify the data’s contents. The only requirement is that both gtrid and
bqual, taken together, must be globally unique.

**Review by ochaloup — COMMENTED** [review]
@ivassile thanks for the enhancements. It seems fine to me.
Just we need to fix the failing test (the `CORE` one, the `AS_TESTS` failure is an intermittent on Narayana CI). And it would be good to create a new test for this behaviour as suggested.

**Comment by mmusgrov** [comment]
> @ivassile I'm sorry for changing my opinion but I verified the XA spec and JTA spec and I think my consideration was wrong last time.
> The spec does not require the global part of `Xid` has not to be null. A Xid in such format should not be removed from recovery processing.
> 
> Details from XA spec (https://pubs.opengroup.org/onlinepubs/009680699/toc.pdf)
> 
> ```
> Although "xa.h" constrains the length and byte alignment of the data element within an
> XID, it does not specify the data’s contents. The only requirement is that both gtrid and
> bqual, taken together, must be globally unique.
> ```
> 
> I would ask you for changing this PR in this regard - ie. do not remove the `xares` when global id is `null` from the `trans` array. Please, change only the logging machinery to not throwing the NPE when it's null (either global, or branch, or format).
> (such change will simultaneously not cause failing the `XARecoveryModuleUnitTest.testXAResourceRecoveryHelperDeregisterLocking`)

@ochaloup is correct, the spec only requires that the Xid be unique. @ivassile can you change the code to only check for zero length bytes in the Xid and if it length is zero then log a warning if the formaId is not equal to -1 (and if it equals -1 then maybe log an info message). It might also be worthwhile logging an info message if either of the gtrid_length or bqual_length vaues are zero.

**Comment by mmusgrov** [comment]
And I think we will need separate tests for `gtrid_length` being zero and `bqual_length` being zero and both being zero,

**Comment by ivassile** [comment]
@mmusgrov Implemented the requested changes. Let me know if this is correct.

File: ArjunaJTA/jta/classes/com/arjuna/ats/jta/xa/XidImple.java [hunk 6]
```diff
@@ -186,15 +187,17 @@ public final void copy(Xid xid) {
...
+				final int gtxlength = (gtx == null ? 0 : gtx.length);
 				final int bqlength = (bql == null ? 0 : bql.length);
 
-				_theXid.gtrid_length = gtx.length;
+				_theXid.gtrid_length = gtxlength;
 				_theXid.bqual_length = bqlength;
 
-				System.arraycopy(gtx, 0, _theXid.data, 0, gtx.length);
+				if (gtxlength > 0) {
+		              System.arraycopy(gtx, 0, _theXid.data, 0, gtxlength);
```
Comment (ochaloup) [inline ArjunaJTA/jta/classes/com/arjuna/ats/jta/xa/XidImple.java:197]: It seems a strange indent is used here.

**Review by ochaloup — COMMENTED** [review]
To me the changes seem fine. I would like just to mention that there is a strange indent formatting at some lines. That could be adjusted.

File: ArjunaJTA/jta/classes/com/arjuna/ats/jta/logging/jtaI18NLogger.java [hunk 7]
```diff
@@ -600,6 +600,14 @@
     @Message(id = 16148, value = "Cannot work further as the argument Xid is null.", format = MESSAGE_FORMAT)
     String get_error_xid_is_null();
 
+    @Message(id = 16149, value = "Returned global transaction identifier and branch qualifier are null but format id is not -1. {0}", format = MESSAGE_FORMAT)
+    @LogMessage(level = WARN)
+    public void warn_recovery_transaction_id_and_branch_qualifier_are_null_wrong_format_id(String arg0);
```
[the remark this answers is not in the document]
Comment (ivassile) [reply ArjunaJTA/jta/classes/com/arjuna/ats/jta/logging/jtaI18NLogger.java]: Corrected

File: ArjunaJTA/jta/classes/com/arjuna/ats/jta/xa/XidImple.java (hunk 6 above)
Comment (ivassile) [reply ArjunaJTA/jta/classes/com/arjuna/ats/jta/xa/XidImple.java]: Corrected

**Review by ochaloup — APPROVED** [review]
@ivassile thanks. It seems fine to me.

...oh, just there was merged another PR meanwhile (ie. https://github.com/jbosstm/narayana/pull/1759) which touches (but only party) the same issue

**Review by mmusgrov — APPROVED** [review]

**Comment by mmusgrov** [comment]
> @ivassile thanks. It seems fine to me.
> 
> ...oh, just there was merged another PR meanwhile (ie. #1759) which touches (but only party) the same issue. A rebase would be needed.

I resolved the conflicts before I saw your comment.
This PR is now good to merge so I will do that now.

**Comment by mmusgrov** [comment]
Ah git is saying it cannot automatically rebase so you will need to do what Ondra suggested, something like
- git pull pr/1747
- git pull --rebase upstream master # where upstream is our repo (upstream ssh://git@github.com/jbosstm/narayana.git)
- fix the the merge conflict (git add, git rebase --continue)
- push to your own topic branch (ivassile:JBTM-3383)
and then we can merge your PR

thanks

**Comment by ivassile** [comment]
@mmusgrov The PR is now rebased and can be merged. This change overrides the change from https://github.com/jbosstm/narayana/pull/1759 and resolves more NPE issues in the code.

**Comment by mmusgrov** [comment]
@ivassile Before we merge this one I have asked @ochaloup why we have two issues and two PRs for the same problem (and it is unclear to me why we need both fixes).

**Comment by ochaloup** [comment]
It seems it's ok I'm merging it. If there is some more remarks we can fix it in a follow-up PR, I assume. Thanks @ivassile for the fix.

**Comment by mmusgrov** [comment]
@ochaloup see my other comments on this

**Comment by mmusgrov** [comment]
It is best not to ignore PR comments before merging

**Comment by mmusgrov** [comment]
> @ivassile Before we merge this one I have asked @ochaloup why we have two issues and two PRs for the same problem (and it is unclear to me why we need both fixes).

This is not a good approach - we will end up with 3 PRs and 3 issues for an non functional issue so we need to get to the bottom of it to avoid it happening again.

**Comment by mmusgrov** [comment]
> It seems it's ok I'm merging it. If there is some more remarks we can fix it in a follow-up PR, I assume. Thanks @ivassile for the fix.

It seems it's ok I'm merging it. If there is some more remarks we can fix it in a follow-up PR, I assume. Thanks @ivassile for the fix.
